File: Core/Specifications/Specification.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;
using Core.Contracts;
using Core.Entities;

namespace Core.Specifications
{
    public class Specification<T> : ISpecification<T>
    {
        protected Func<T, bool> _filter;
        public Specification(IFilter<T> filter)
        {            
            _filter = filter.ToPredicate();
        }

        public Specification(Func<T, bool> predicate)
        {
            _filter = predicate;
        }

        public bool isSatisfiedBy(T entity)
        {
            return _filter(entity);
        }

        public ISpecification<T> And(ISpecification<T> criteria) 
        {
            return new AndSpecification<T>(this, criteria);
        }

        public ISpecification<T> Or(ISpecification<T> criteria) 
        {
            return new OrSpecification<T>(this, criteria);
        }

        public ISpecification<T> Not() 
        {
            return new NotSpecification<T>(this);
        }

        public IQueryable<T> AsQueryable(IQueryable<T> query) 
        {
            return query.Where(e => isSatisfiedBy(e));
        }
    }
}

File: Client/projects/core/src/lib/contracts/specifications/Specification.ts
import { ISpecification } from "./";


export abstract class Specification<T> implements ISpecification<T> {    
    constructor() {
        this.isSatisfiedBy = this.isSatisfiedBy.bind(this);
    }

    abstract isSatisfiedBy(target: T): boolean;

    or(specification: ISpecification<T>): Specification<T> {
        return new OrSpecification<T>(this, specification);
    }

    and(specification: ISpecification<T>): Specification<T> {
        return new AndSpecification<T>(this, specification);
    }

    not(): Specification<T> {
        return new NotSpecification<T>(this);
    }
}


export abstract class CompositeSpecification<T> extends Specification<T> {
    constructor(protected left: ISpecification<T>, protected right: ISpecification<T>) {
        super();
    }
}


export class AndSpecification<T> extends CompositeSpecification<T> {
    constructor(left: ISpecification<T>, right: ISpecification<T>) {
        super(left, right);
    }

    isSatisfiedBy(target: T): boolean {
        return this.left.isSatisfiedBy(target) && this.right.isSatisfiedBy(target);
    }
}


export class OrSpecification<T> extends CompositeSpecification<T> {
    constructor(left: ISpecification<T>, right: ISpecification<T>) {
        super(left, right);
    }

    isSatisfiedBy(target: T): boolean {
        return this.left.isSatisfiedBy(target) || this.right.isSatisfiedBy(target);
    }
}


export class NotSpecification<T> extends Specification<T> {
    constructor(protected specification: ISpecification<T>) {  
        super();
    }

    isSatisfiedBy(target: T): boolean {
        return !this.specification.isSatisfiedBy(target);
    }
}